rust/cyclotron-fetch/src/fetch.rs

```diff
@@ -623,52 +619,53 @@ where
 // Complete the job with some result.
 pub async fn complete_job(
     worker: &Worker,
-    job: &Job,
+    job_id: Uuid,
     return_queue: String,
     on_finish: OnFinish,
     result: FetchResult,
-    body: Option<Bytes>,
 ) -> Result<(), FetchError> {
-    worker.set_state(job.id, JobState::Available)?;
-    worker.set_queue(job.id, &return_queue)?;
+    worker.set_state(job_id, JobState::Available)?;
+    worker.set_queue(job_id, &return_queue)?;
+    let (result, body) = result.take_body();
 
     let is_success = result.is_success();
 
-    let result = do_or_dead_letter(worker, job.id, || serde_json::to_string(&result)).await??;
+    let result = do_or_dead_letter(worker, job_id, || serde_json::to_vec(&result)).await??;
 
     match (on_finish, is_success) {
         (OnFinish::Complete, true) => {
-            worker.set_state(job.id, JobState::Completed)?;
+            worker.set_state(job_id, JobState::Completed)?;
         }
         (OnFinish::Complete, false) => {
-            worker.set_state(job.id, JobState::Failed)?;
+            worker.set_state(job_id, JobState::Failed)?;
         }
         (OnFinish::Return, _) => {
             // If we're retuning the job, we don't care whether it succeeded or not, the caller wants it back
-            worker.set_state(job.id, JobState::Available)?;
+            worker.set_state(job_id, JobState::Available)?;
         }
     }
 
-    worker.set_parameters(job.id, Some(result))?;
-    worker.set_blob(job.id, body)?;
-    worker.set_metadata(job.id, None)?; // We're finished with the job, so clear our internal state
-    worker.flush_job(job.id).await?;
+    worker.set_priority(job_id, 0)?; // Reset job priority on completion
```

Comment:
Are you sure? I thought QoS stuff might adjust a jobs priority based on the team/function it came from, I'm not sure why finishing the fetch and sending it back to Hog (for example) should necessarily always hardcode the priority to 0?

Reply:
The QoS override stuff will be calculated as deltas from the "base" priority in the job row... and yeah I'm not certain to be honest. The two arguments in my head are:
- Once we finish the fetch, this hog function is exactly as executable and likely to succeed as any other one from what I understand
- If this function took a bunch of retries to proceed, maybe we DO want to spend less compute on it if there's a less flaky function we could be running instead?

I'm leaning towards "reset priority for all jobs where the fetch succeed no matter how many tries it took, but if the final outcome is failure, leave it down-prioritised", what do you think?

Reply:
I think Ben has his own reasons for priorities. Or maybe priorities is just a confusing word we've chosen to use here. CDP already has the idea of some "baseline" then a "degraded" and "disabled" state, and also "overflow" (not sure if one of those is a dupe). So there might be different consumer/worker deployments running to process different qualities work (based on team/function/etc). That's why it didn't make sense to me to reset anything after fetch is done -- I think he expects work to remain in the same "lane" unless we send a config-plane request to reschedule all of the jobs.

As usual though, this is underspecified. 🤷‍♂️

Reply:
We should discuss, because CDP already has function states, an API to retrieve them, etc that is per Hog Function.

https://github.com/PostHog/posthog/blob/af509c804aba3232e686eb5f7382e4108d64f74d/posthog/models/hog_functions/hog_function.py#L25-L31

https://github.com/PostHog/posthog/blob/af509c804aba3232e686eb5f7382e4108d64f74d/posthog/models/hog_functions/hog_function.py#L69-L88

https://github.com/PostHog/posthog/blob/af509c804aba3232e686eb5f7382e4108d64f74d/plugin-server/src/cdp/cdp-api.ts#L42-L45

Reply:
And maybe we need both ideas, but the question then is whether both ideas need to be first class citizens inside of Cyclotron?